kubernetes/samples/scripts/domain-lifecycle/helper.sh

```diff
@@ -249,13 +249,56 @@ function createPatchJsonToUpdateClusterPolicy {
   local policy=$3
   local __result=$4
   
-  replacePolicyCmd="(.spec.clusters[] | select (.clusterName == \"${clusterName}\") \
-    | .serverStartPolicy) |= \"${policy}\""
-  startPolicy=$(echo ${domainJson} | jq "${replacePolicyCmd}" | jq -cr '(.spec.clusters)')
+  clusters=$(echo ${domainJson} | jq -cr '(.spec.clusters)')
+  if [ "${clusters}" == "null" ]; then
+    # cluster doesn't exist, add cluster with server start policy
+    addClusterStartPolicyCmd=".[.| length] |= . + {\"clusterName\":\"${clusterName}\", \
+      \"serverStartPolicy\":\"${policy}\"}"
+    startPolicy=$(echo ${clusters} | jq -c "${addClusterStartPolicyCmd}")
+  else
+    mapCmd="\
+      . |= (map(.clusterName) | index (\"${clusterName}\")) as \$idx | \
+      if \$idx then \
+      .[\$idx][\"serverStartPolicy\"] = \"${policy}\" \
+      else .+  [{clusterName: \"${clusterName}\" , serverStartPolicy: \"${policy}\"}] end"
+    startPolicy=$(echo ${clusters} | jq "${mapCmd}")
+  fi
+
   patchJson="{\"spec\": {\"clusters\": "${startPolicy}"}}"
   eval $__result="'${patchJson}'"
 }
 
+#
+# Function to create patch json to update cluster replicas
+# $1 - Domain resource in json format
+# $2 - Name of cluster whose replicas will be patched
+# $3 - replica count
+# $4 - Return value containing patch json string
+#
+function createPatchJsonToUpdateReplicas {
+  local domainJson=$1
+  local clusterName=$2
+  local replicas=$3
+  local __result=$4
+
+  clusters=$(echo ${domainJson} | jq -cr '(.spec.clusters)')
+  if [ "${clusters}" == "null" ]; then
+    # cluster doesn't exist, add cluster with replicas
+    addClusterReplicasCmd=".[.| length] |= . + {\"clusterName\":\"${clusterName}\", \
+      \"replicas\":${replicas}}"
+    replicasPatch=$(echo ${clusters} | jq -c "${addClusterReplicasCmd}")
+  else
+    mapCmd="\
+      . |= (map(.clusterName) | index (\"${clusterName}\")) as \$idx | \
+      if \$idx then \
+      .[\$idx][\"replicas\"] = ${replicas} \
+      else .+  [{clusterName: \"${clusterName}\" , replicas: ${replicas}}] end"
+    replicasPatch=$(echo ${clusters} | jq "${mapCmd}")
+  fi
+  patchJson="{\"spec\": {\"clusters\": "${replicasPatch}"}}"
```

Comment:
addClustersReplicasCmd, mapCmd, replicasPath, and patchJson should all be declared as local.  They also should be sufficiently uniquely named so that it's unlikely their names can collide with the value that might be in "$__result".

overall, it might be good overall pattern to have all 'local' variables start with __, and then avoid using __ for global vars...

Reply:
I'm using the __ pattern for the return parameters. For now, I'll make sure that local parameters are not used in the global scope.

Reply:
Thinking about this some more, I assume they occlude if they're declared local - which would make them safe...

Reply:
... as long as they're declared local...